x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/persistence/AnomalyDetectorsIndex.java

```diff
@@ -64,4 +82,66 @@ public static String configIndexName() {
         return AnomalyDetectorsIndexFields.CONFIG_INDEX;
     }
 
+    /**
+     * Create the .ml-state index (if necessary)
+     * Create the .ml-state-write alias for the .ml-state index (if necessary)
+     */
+    public static void createStateIndexAndAliasIfNecessary(Client client, ClusterState state, final ActionListener<Boolean> finalListener) {
+
+        if (state.getMetaData().getAliasAndIndexLookup().containsKey(jobStateIndexWriteAlias())) {
+            finalListener.onResponse(false);
+            return;
+        }
+
+        final ActionListener<String> createAliasListener = ActionListener.wrap(
+            concreteIndexName -> {
+                final IndicesAliasesRequest request = client.admin()
+                    .indices()
+                    .prepareAliases()
+                    .addAlias(concreteIndexName, jobStateIndexWriteAlias())
+                    .request();
+                executeAsyncWithOrigin(client.threadPool().getThreadContext(),
+                    ML_ORIGIN,
+                    request,
+                    ActionListener.<AcknowledgedResponse>wrap(
+                        resp -> finalListener.onResponse(resp.isAcknowledged()),
+                        finalListener::onFailure),
+                    client.admin().indices()::aliases);
+            },
+            finalListener::onFailure
+        );
+
+        IndexNameExpressionResolver indexNameExpressionResolver = new IndexNameExpressionResolver();
+        String[] state_indices = indexNameExpressionResolver.concreteIndexNames(state,
+            IndicesOptions.lenientExpandOpen(),
+            jobStateIndexPattern());
+        if (state_indices.length > 0) {
+            List<String> indices = Arrays.asList(state_indices);
+            indices.sort(String::compareTo);
+            createAliasListener.onResponse(indices.get(indices.size() - 1));
```

Comment:
Instead of creating the temporary list just for sorting you could sort the array directly:

```
Arrays.sort(stateIndices);
createAliasListener.onResponse(stateIndices[stateIndices.length - 1]);
```

or:

```
Arrays.sort(stateIndices, Collections.reverseOrder());
createAliasListener.onResponse(stateIndices[0]);
```

or:

```
createAliasListener.onResponse(Arrays.stream(stateIndices).max(String::compareTo).get());
```

Reply:
Lulz, can't believe I missed that.